src/basic/depgraphtypes.nim

```diff
@@ -1,183 +1,111 @@
-  result = (dest, if dirExists(dest): DoNothing else: DoClone)
-
-proc toDestDir*(g: DepGraph; d: Dependency): string =
+import std / [paths, tables, streams, json, jsonutils]
+
+import sattypes, context, deptypes, reporters, nimblecontext, pkgurls, versions
+
+
+type 
+  VisitState = enum
+    NotVisited, InProgress, Visited
+
+proc toposorted*(graph: DepGraph): seq[Package] =
+  ## Returns a sequence of packages in topological order
+  ## Packages that are depended upon come before packages that depend on them
+  result = @[]
+  var visited = initTable[PkgUrl, VisitState]()
+  
+  # Initialize all packages as not visited
+  for url, pkg in graph.pkgs:
+    visited[url] = NotVisited
+  
+  # DFS-based topological sort
+  proc visit(pkg: Package): seq[Package] =
+    if visited[pkg.url] == Visited:
+      return
+    if visited[pkg.url] == InProgress:
+      # This means we have a cycle, which shouldn't happen in a valid dependency graph
+      # But we'll handle it gracefully
+      return
+    
+    visited[pkg.url] = InProgress
+    
+    # Get the active release to check its dependencies
+    let release = pkg.activeNimbleRelease()
+    if not release.isNil:
+      # Visit all dependencies first
+      for (depUrl, _) in release.requirements:
+        if depUrl in graph.pkgs:
+          let depPkg = graph.pkgs[depUrl]
+          result.add visit(depPkg)
+    
+    # Mark as visited and add to result
+    visited[pkg.url] = Visited
+    result.add(pkg)
+  
+  # Start with root package
+  if not graph.root.isNil:
+    result.add visit(graph.root)
+  
+  # Visit any remaining packages (disconnected or not reachable from root)
+  for url, pkg in graph.pkgs:
+    if visited[url] == NotVisited:
+      result.add visit(pkg)
+
+proc validateDependencyGraph*(graph: DepGraph): bool =
+  ## Checks if the dependency graph is valid (no cycles)
+  var visited = initTable[PkgUrl, VisitState]()
+  
+  # Initialize all packages as not visited
+  for url, pkg in graph.pkgs:
+    visited[url] = NotVisited
+  
+  proc checkCycles(pkg: Package): bool =
+    if visited[pkg.url] == Visited:
+      return true
+    if visited[pkg.url] == InProgress:
+      # Cycle detected
+      return false
+    
+    visited[pkg.url] = InProgress
+    
+    # Check all dependencies
+    let release = pkg.activeNimbleRelease()
+    if not release.isNil:
+      for (depUrl, _) in release.requirements:
+        if depUrl in graph.pkgs:
+          let depPkg = graph.pkgs[depUrl]
+          if not checkCycles(depPkg):
+            return false
+    
+    visited[pkg.url] = Visited
+    return true
+  
+  # Check from all possible starting points
+  for url, pkg in graph.pkgs:
+    if visited[url] == NotVisited:
+      if not checkCycles(pkg):
+        return false
+  
+  return true
+
+proc toDestDir*(g: DepGraph; d: Package): Path =
   result = d.ondisk
 
-proc enrichVersionsViaExplicitHash*(versions: var seq[DependencyVersion]; x: VersionInterval) =
```

Comment:
How come this logic isn't required anymore?

Reply:
The logic is there, it's now handled in an additional pass at the end of `expandGraph` by calling `traverseDependency` with an `ExplicitVersions` traverse mode. That way all the explicit versions in all requirements are processed.